cmake/DaemonArchitecture.cmake

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) 2022, Daemon Developers
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of the <organization> nor the
+#    names of its contributors may be used to endorse or promote products
+#    derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+################################################################################
+# Determine architecture
+################################################################################
+
+# When adding a new architecture, look at all the places ARCH is used
+
+if(DAEMON_DIR)
+	set(TARGET_DETECT_SOURCE_FILE "${DAEMON_DIR}/src/common/Architecture.cpp")
+else()
+	set(TARGET_DETECT_SOURCE_FILE "${CMAKE_CURRENT_SOURCE_DIR}/src/common/Architecture.cpp")
+endif()
+
+try_run(
+	run_result_unused
+	compile_result_unused
+	"${CMAKE_BINARY_DIR}"
+	"${TARGET_DETECT_SOURCE_FILE}"
+	CMAKE_FLAGS CMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES}
+	COMPILE_DEFINITIONS -DPREPROCESSOR_ARCH_DETECTION=1
+	COMPILE_OUTPUT_VARIABLE ARCH
+)
+
+string(REGEX MATCH "DAEMON_ARCH ([a-zA-Z0-9_]+)" ARCH "${ARCH}")
+string(REPLACE "DAEMON_ARCH " "" ARCH "${ARCH}")
+
+if (NOT ARCH)
+	set(ARCH "unsupported")
+endif()
+
+if (ARCH STREQUAL "unsupported")
+	message(FATAL_ERROR "Architecture not supported")
+endif()
+
+add_definitions(-DDAEMON_ARCH_${ARCH})
+
+# Compatibility with historic NACL_ARCH suffixes for .nexe file, modifying
+# them breaks engine compatibility with published nexe game binaries.
+# The string is also used in some logs.
+
+if (ARCH STREQUAL "i686")
+	set(ARCH_STRING "x86")
+elseif(ARCH STREQUAL "amd64")
+	set(ARCH_STRING "x86_64")
+else()
+	set(ARCH_STRING "${ARCH}")
+endif()
+
+# Quotes are part of the define.
```

Comment:
It would be good to actually do this. Like `-DARCH_STRING="\"${ARCH_STRING}\""` I think. Then you won't have to add STRING macros everywhere it is used.

Reply:
I tried that, the contradictory “Quotes are part of the define.” comment is a relic of that (I should have removed this comment). This doesn't work, sometime it ends as `"<arch>"`, sometime it ends as  `""<arch>"", and both variants ends in the same PCH. Here it's written that quote characters aren't reliable anyway:

- https://cmake.org/cmake/help/latest/prop_dir/COMPILE_DEFINITIONS.html

That'd why I went back to `STRING` macros.

Reply:
OK. You should use STRING instead of XSTRING; XSTRING is for when the argument is a macro and you want to expand it first before stringizing the expanded contents

Reply:
Done.

Reply:
I had to revert to `XSTRING` because otherwise I would get that:

```
Warn: VM module file not found: sgame-NACL_ARCH_STRING.nexe 
Warn: NaCl integrated runtime not found: irt_core-NACL_ARCH_STRING.nexe 
Loading VM module sgame-NACL_ARCH_STRING.nexe... 
Using loader args:  nacl_helper_bootstrap nacl_loader --r_debug=0xXXXXXXXXXXXXXXXX --reserved_at_zero=0xXXXXXXXXXXXXXXXX -v -B irt_core-NACL_ARCH_STRING.nexe -e -i 100:25 -- sgame-NACL_ARCH_STRING.nexe 100 
Warn: IPC: Socket closed by remote end 
```